### ISSUE TITLE
Update AWS Provisioner

### ISSUE THREAD

**Issue by mafernando** [projectjellyfish/jellyfish-fog#17]
- [x] Add the following attributes to AWS provisioning questions
  - [x] VPC
  - [x] Subnet
  - [x] Security Group
- [x] Add ability to provision more than t1.micros
  - Solution: Need to specify `flavor_id` attribute instead  `instance_id` for size of instance
- [x] Add product types for the following 
  - [x] VPC
  - [x] Subnet
  - [x] Security Group
- [x] Add ability to add EBS storage (maybe as a new product type?)
- [x] Cut new jellyfish-fog gem
  - [x] Verify AWS provisioning of all product types
  -  [x] Push updated gem to RubyGems 
  - [x] Update API lockfile with reference to update gem
- ~~Need to decide what kinds of options we want to provide beyond the basic for the newly added product types~~ - _to be revisited later_

**Comment by mafernando** [comment]
**References:**
- ami-0bc9d462 [Community CentOS6.5_HVM_SR-IOV_ixgbevf-2.14.2 AMI](http://geekblood.com/2014/12/17/chef-12-jenkins-and-aws-provisioning-part-1)
- ami-1ccae774 [Amazon Linux AMI 2015.03.0 x86_64 PV EBS](http://thecloudmarket.com/image/ami-1ccae774--amzn-ami-pv-2015-03-0-x86-64-ebs)
- [security group attributes fog aws](https://github.com/mitchellh/vagrant-aws/blob/d77a8b94e24fe3e7864a74634d99b5b1afbf2660/lib/vagrant-aws/action/run_instance.rb#L95-L99)
- [how to add existing ebs volumes](https://groups.google.com/forum/#!topic/ruby-fog/uLDZ2hTdJ0I)
- [fog compute vpc model v1](https://github.com/fog/fog-aws/blob/5903a7819b12ec6feb3b1d4177f9a1f4a4ca3a59/lib/fog/aws/models/compute/vpcs.rb)
- [fog compue vpc model v2](http://www.rubydoc.info/github/fog/fog/Fog/Compute/AWS/Real:create_vpc)
- [convert json to ruby hash](http://stackoverflow.com/questions/7964282/convert-json-to-ruby-hash)
- [create subnet fog](http://www.rubydoc.info/github/fog/fog/Fog/Compute/AWS/Real#create_subnet-instance_method)
- [create security subgroup](http://www.rubydoc.info/github/fog/fog/Fog/Compute/AWS/Real#create_security_group-instance_method)
- [network vs broadcast id & cidr](https://www.countryipblocks.net/identifying-the-network-and-broadcast-address-of-a-subnet)
- [create ebs with fog](https://gist.github.com/724912/116f35d0b7ab30db765b858faea123919592d067)
- [delete rds with fog](http://www.rubydoc.info/github/fog/fog/Fog/AWS/RDS/Real#delete_db_instance-instance_method)

**Comment by mafernando** [comment]
**Issues:**
- configuration details may need to be revisited back in jellyfish api. the issue here is that default options in `details` need to be `nil` not empty string. See AWS Infrastructure provisioner.
- not sure if `connection.delete_vpc` will remove all associated resources with the VPC, may want to look into this
- the object mocked in specs does not match the object generated for Fog corresponding to a provision response. this causes problems with accessing attributes using symbols or strings for keys.
  - _solution: since mock object doesn't serialize the payload_response as JSON, manually add that step in before retirement spec_
  - [x] follow-up: need to revisit AWS Infrastructure provisioning and make sure that the payload is stored correctly. When the response is converted to JSON, the any symbol keys contained therein will be flattened and will need to be referenced using strings.

**Comment by mafernando** [comment]
**Examples:**
- create a vpc on AWS (see method [here](http://www.rubydoc.info/github/fog/fog/Fog/Compute/AWS/Real:create_vpc))
  
  ```
  [20] pry(#<Jellyfish::Fog::AWS::VPC>)> connection.create_vpc(details['cidr_block'])
  => #<Excon::Response:0x007fe6767abe08
  @body="",
  @data=
  {:body=>{"vpcSet"=>[{"tagSet"=>{}, "vpcId"=>"vpc-b8053cdd", "state"=>"pending", "cidrBlock"=>"10.0.0.0/16", "dhcpOptionsId"=>"dopt-0ce6fc6e"}], "requestId"=>"XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"},
  :headers=>{"Content-Type"=>"text/xml;charset=UTF-8", "Vary"=>"Accept-Encoding", "Date"=>"Mon, 01 Jun 2015 19:57:53 GMT", "Server"=>"AmazonEC2"},
  :status=>200,
  :status_line=>"HTTP/1.1 200 OK\r\n",
  :reason_phrase=>"OK",
  :remote_ip=>"ZZZ.ZZZ.ZZZ.Z",
  :local_port=>#####,
  :local_address=>"YYY.YYY.YYY.YYY"},
  @headers={"Content-Type"=>"text/xml;charset=UTF-8", "Vary"=>"Accept-Encoding", "Date"=>"Mon, 01 Jun 2015 19:57:53 GMT", "Server"=>"AmazonEC2"},
  @local_address="YYY.YYY.YYY.YYY",
  @local_port=60972,
  @remote_ip="ZZZ.ZZZ.ZZZ.Z",
  @status=200>
  [21] pry(#<Jellyfish::Fog::AWS::VPC>)> connection.vpcs
  => [  <Fog::Compute::AWS::VPC
  id="vpc-XXXXXXXX",
  state="available",
  cidr_block="10.0.0.0/16",
  dhcp_options_id="dopt-0ce6fc6e",
  tags={},
  tenancy="default"
  >]
  ```

**Comment by mafernando** [comment]
Closed with 03b778e5877aed42f8ccdbe2e656b3ab089d8bf6
